### PR TITLE
Write a tutorial for dates, schedules, calendars and dcf

### PR DESCRIPTION
**Pull request by markus-da**
Fixes https://github.com/digital-asset/daml-finance/issues/1044

### CODE REVIEW

**Comment by matteolimberto-da** [comment]
Do we think this belongs to advanced topics? There is almost no pre-requisite knowledge required, so I would see it belong to a different section

File: docs/2.8.0/docs/daml-finance/tutorials/advanced-topics/instrument-modeling/calendar-dcf.rst [hunk 1]
```diff
@@ -0,0 +1,268 @@
+.. Copyright (c) 2023 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+.. SPDX-License-Identifier: Apache-2.0
```
Comment (matteolimberto-da) [inline docs/2.8.0/docs/daml-finance/tutorials/advanced-topics/instrument-modeling/calendar-dcf.rst:2]: rename the file to `date-utilities`?

File: docs/2.8.0/docs/daml-finance/tutorials/advanced-topics/instrument-modeling/calendar-dcf.rst [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+
+How to use the calendar, schedule and day count utility functions
+#################################################################
+
+The Daml Finance library contains date related utility functions that are used to implement industry
+standard conventions. These are mainly used internally in instruments like Bonds and Swaps. However,
+there are use cases where it would be helfpul for users to understand these utility functions in
```
Comment (matteolimberto-da) [inline docs/2.8.0/docs/daml-finance/tutorials/advanced-topics/instrument-modeling/calendar-dcf.rst:9]:
```suggestion
there are use cases where it would be helpful for users to understand these utility functions in
```

File: docs/2.8.0/docs/daml-finance/tutorials/advanced-topics/instrument-modeling/calendar-dcf.rst [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+more detail, for example if you want to:
```
Comment (matteolimberto-da) [inline docs/2.8.0/docs/daml-finance/tutorials/advanced-topics/instrument-modeling/calendar-dcf.rst:10]:
```suggestion
more detail. For example if you want to:
```

File: docs/2.8.0/docs/daml-finance/tutorials/advanced-topics/instrument-modeling/calendar-dcf.rst [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+
+- develop your own instruments that depend on business day shifts, schedules, and/or day count
+  conventions
+- understand and validate the implementation of current Daml Finance instruments
```
Comment (matteolimberto-da) [inline docs/2.8.0/docs/daml-finance/tutorials/advanced-topics/instrument-modeling/calendar-dcf.rst:14]:
```suggestion
- understand and validate the implementation of existing Daml Finance instruments
```

File: docs/2.8.0/docs/daml-finance/tutorials/advanced-topics/instrument-modeling/calendar-dcf.rst [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+- compose existing Daml Finance instruments, for example define a series of zero coupon bonds
+  according to a specific schedule of expiries
+- use this calendar and schedule related functionality for non-financial use cases
```
Comment (matteolimberto-da) [inline docs/2.8.0/docs/daml-finance/tutorials/advanced-topics/instrument-modeling/calendar-dcf.rst:17]:
```suggestion
- use these calendar and schedule related functionalities for non-financial use cases
```

File: docs/2.8.0/docs/daml-finance/tutorials/advanced-topics/instrument-modeling/calendar-dcf.rst [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+Business days and non-business days are distinguished by defining a holiday calendar, for example:
+
+.. literalinclude:: ../../../src/test/daml/Daml/Finance/Util/Test/Date/Calendar.daml
+  :language: daml
+  :start-after: -- CREATE_HOLIDAY_CALENDAR_DATA_BEGIN
+  :end-before: -- CREATE_HOLIDAY_CALENDAR_DATA_END
+
+The :ref:`HolidayCalendarData <type-daml-finance-interface-types-date-calendar-holidaycalendardata-60004>`
+type describes the set of non-business days by defining which days of the week belong to the weekend
+and which dates are general holidays.
```
Comment (matteolimberto-da) [inline docs/2.8.0/docs/daml-finance/tutorials/advanced-topics/instrument-modeling/calendar-dcf.rst:34]:
```suggestion
and which dates are additional holidays.
```

File: docs/2.8.0/docs/daml-finance/tutorials/advanced-topics/instrument-modeling/calendar-dcf.rst [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+utility functions to add a period to a given date.
+
+For example, you can add a day, week, month or a year to a given date:
+
+.. literalinclude:: ../../../src/test/daml/Daml/Finance/Util/Test/Date/Calendar.daml
+  :language: daml
+  :start-after: -- TEST_ADD_PERIOD_DAY_WEEK_MONTH_YEAR_BEGIN
+  :end-before: -- TEST_ADD_PERIOD_DAY_WEEK_MONTH_YEAR_END
+
+This function can also handle edge cases, for example in case of a monthly roll to a date that may
```
Comment (matteolimberto-da) [inline docs/2.8.0/docs/daml-finance/tutorials/advanced-topics/instrument-modeling/calendar-dcf.rst:95]: We haven't defined what a "roll" is

File: docs/2.8.0/docs/daml-finance/tutorials/advanced-topics/instrument-modeling/calendar-dcf.rst [hunk 8]
```diff
@@ -0,0 +1,268 @@
...
+.. literalinclude:: ../../../src/test/daml/Daml/Finance/Util/Test/Date/Calendar.daml
+  :language: daml
+  :start-after: -- TEST_PREVIOUS_DOM_FROM_EOFEB_BEGIN
+  :end-before: -- TEST_PREVIOUS_DOM_FROM_EOFEB_END
+
+Schedule
+========
+
+While the above :ref:`RollConvention <module-daml-finance-util-date-rollconvention-88672>` functions
+can be useful on their own, they are often required in the context of rolling out a schedule. This
```
Comment (matteolimberto-da) [inline docs/2.8.0/docs/daml-finance/tutorials/advanced-topics/instrument-modeling/calendar-dcf.rst:152]: I think we should define what a schedule is

Comment (matteolimberto-da) [reply docs/2.8.0/docs/daml-finance/tutorials/advanced-topics/instrument-modeling/calendar-dcf.rst:152]: Might be worth explaining unadjusted vs adjusted dates

File: docs/2.8.0/docs/daml-finance/tutorials/advanced-topics/instrument-modeling/calendar-dcf.rst [hunk 9]
```diff
@@ -0,0 +1,268 @@
...
+DayCount
+========
+
+Many instruments that pay interest require an exact definition of how many days of interest belong
+to each payment period. The :ref:`DayCount <module-daml-finance-util-date-daycount-38239>` module
+provides functions for this. In particular, a *day count fraction* (or *dcf* in short) can be
+calculated between two dates. This indicates the fraction of a full year.
+
+For example, consider a one year bond that pays a quarterly coupon according to a given schedule,
+like the one described in the previous chapter.
```
Comment (matteolimberto-da) [inline docs/2.8.0/docs/daml-finance/tutorials/advanced-topics/instrument-modeling/calendar-dcf.rst:223]:
```suggestion
like the one described in the previous section.
```

File: docs/2.8.0/docs/daml-finance/tutorials/advanced-topics/instrument-modeling/calendar-dcf.rst [hunk 10]
```diff
@@ -0,0 +1,268 @@
...
+
+The schedule can be used to represent different aspects of a financial instrument, for example:
+
+- payment periods, which define e.g. *when* a bond coupon or a swap payment should be paid
+- calculation periods, which define e.g. *how* the interest amount is calculated
+
+DayCount
+========
+
+Many instruments that pay interest require an exact definition of how many days of interest belong
```
Comment (matteolimberto-da) [inline docs/2.8.0/docs/daml-finance/tutorials/advanced-topics/instrument-modeling/calendar-dcf.rst:217]: We should specify that this is relevant as the coupon rate is often expressed in annualized terms

**Review by matteolimberto-da — COMMENTED** [review]
Thanks, I left a few comments, let me know what you think

**Comment by markus-da** [comment]
> Do we think this belongs to advanced topics? There is almost no pre-requisite knowledge required, so I would see it belong to a different section

I thought about that, and I think it depends. The use cases mentioned in the beginning seem to indicate rather advanced topics, compared to just using Daml Finance:

- develop your own instruments that depend on business day shifts, schedules, and/or day count
  conventions
- understand and validate the implementation of existing Daml Finance instruments
- compose existing Daml Finance instruments, for example define a series of zero coupon bonds
  according to a specific schedule of expiries
- use these calendar and schedule related functionalities for non-financial use cases

I thought about putting it into a new section Extending Daml Finance, but it only fits with some of those use cases.

What do you think?

File: docs/2.8.0/docs/daml-finance/tutorials/advanced-topics/instrument-modeling/calendar-dcf.rst (hunk 1 above)
Comment (markus-da) [reply docs/2.8.0/docs/daml-finance/tutorials/advanced-topics/instrument-modeling/calendar-dcf.rst]: After I changed the name, GitHub would not let me Commit your suggestions. Applied them manually instead.

**Comment by matteolimberto-da** [comment]
> > Do we think this belongs to advanced topics? There is almost no pre-requisite knowledge required, so I would see it belong to a different section
> 
> I thought about that, and I think it depends. The use cases mentioned in the beginning seem to indicate rather advanced topics, compared to just using Daml Finance:
> 
> * develop your own instruments that depend on business day shifts, schedules, and/or day count
>   conventions
> * understand and validate the implementation of existing Daml Finance instruments
> * compose existing Daml Finance instruments, for example define a series of zero coupon bonds
>   according to a specific schedule of expiries
> * use these calendar and schedule related functionalities for non-financial use cases
> 
> I thought about putting it into a new section Extending Daml Finance, but it only fits with some of those use cases.
> 
> What do you think?

Your argument makes sense 👍 you effectively need this stuff only when implementing new instruments

**Review by matteolimberto-da — APPROVED** [review]

File: docs/2.8.0/docs/daml-finance/tutorials/advanced-topics/instrument-modeling/date-utilities.rst [hunk 11]
```diff
@@ -0,0 +1,279 @@
...
+We can compare this initial day count fraction with those of the 2nd and 3rd periods of this
+schedule, which are both regular (3M):
+
+.. literalinclude:: ../../../src/test/daml/Daml/Finance/Util/Test/Date/DayCount.daml
+  :language: daml
+  :start-after: -- CALCULATE_ACT360_DCF_PERIOD2AND3_BEGIN
+  :end-before: -- CALCULATE_ACT360_DCF_PERIOD2AND3_END
+
+These day count fractions are clearly greater than the one of the first period, since they
+correspond to 3M instead of 2M. Note that they differ slightly, because they do not contain the
```
Comment (marciw-da) [inline docs/2.8.0/docs/daml-finance/tutorials/advanced-topics/instrument-modeling/date-utilities.rst:262]:
```suggestion
correspond to 3M instead of 2M. Note that they differ slightly: they do not contain the
```

File: docs/2.8.0/docs/daml-finance/tutorials/advanced-topics/instrument-modeling/date-utilities.rst [hunk 12]
```diff
@@ -0,0 +1,279 @@
...
+exact same number of days.
+
+In addition to ``Act360``, the
+:ref:`DayCountConventionEnum <type-daml-finance-interface-types-date-daycount-daycountconventionenum-67281>`
+also supports several other day count conventions. Some of them can be computed using
+:ref:`calcDcf <function-daml-finance-util-date-daycount-calcdcf-20432>`, using only a start and an
+end date as an input, as described above.
```
Comment (marciw-da) [inline docs/2.8.0/docs/daml-finance/tutorials/advanced-topics/instrument-modeling/date-utilities.rst:269]:
```suggestion
exact same number of days.

In addition to ``Act360``, the
:ref:`DayCountConventionEnum <type-daml-finance-interface-types-date-daycount-daycountconventionenum-67281>`
supports several other day count conventions. You can compute some of them with
:ref:`calcDcf <function-daml-finance-util-date-daycount-calcdcf-20432>`, using only a start and an
end date as an input as shown above.
```

File: docs/2.8.0/docs/daml-finance/tutorials/advanced-topics/instrument-modeling/date-utilities.rst [hunk 13]
```diff
@@ -0,0 +1,279 @@
...
+Others, e.g. ``ActActISDA``, are a bit more complicated because they require additional information.
+They can be calculated using the
+:ref:`calcPeriodDcf <function-daml-finance-util-date-daycount-calcperioddcf-63067>` function
+instead, which takes a schedule period (containing stub information) as input, as well as the
+schedule end date and the payment frequency:
```
Comment (marciw-da) [inline docs/2.8.0/docs/daml-finance/tutorials/advanced-topics/instrument-modeling/date-utilities.rst:274]:
```suggestion
Others, such as ``ActActISDA``, are a bit more complicated because they require additional information.
You can calculate them using the
:ref:`calcPeriodDcf <function-daml-finance-util-date-daycount-calcperioddcf-63067>` function
instead. This function takes a schedule period (containing stub information) as input, as well as the
schedule end date and the payment frequency:
```

File: docs/2.8.0/docs/daml-finance/tutorials/advanced-topics/intro.rst [hunk 14]
```diff
@@ -15,3 +15,7 @@ The following tutorials are available:
 * The :doc:`Contingent Claims Instrument <instrument-modeling/contingent-claims-instrument>`
   tutorial describes how to leverage the Contingent Claims library to lifecycle custom instrument
   implementations.
+
+* The :doc:`Date utility <instrument-modeling/date-utilities>`
+  tutorial describes how to use the calendar, schedule and day count utility functions, e.g. in
+  order to create your own instruments.
```
Comment (marciw-da) [inline docs/2.8.0/docs/daml-finance/tutorials/advanced-topics/intro.rst:21]:
```suggestion
* The :doc:`Date utility <instrument-modeling/date-utilities>`
  tutorial shows you how to use the calendar, schedule, and day count utility functions (for example, to
  create your own instruments).
```

File: docs/2.8.0/docs/daml-finance/tutorials/advanced-topics/instrument-modeling/date-utilities.rst [hunk 15]
```diff
@@ -0,0 +1,279 @@
...
+Now, let us roll out this schedule to get the specific start and end date for each period, and
+compare it to the expected result:
+
+.. literalinclude:: ../../../src/test/daml/Daml/Finance/Util/Test/Date/Schedule.daml
+  :language: daml
+  :start-after: -- CREATE_SCHEDULE_BEGIN
+  :end-before: -- CREATE_SCHEDULE_END
+
+An important aspect of schedules is the concept of *stub periods*. The above schedule only had
+regular 3M periods. However, if the sample schedule would start one month later:
```
[the remark this answers is not in the document]
Comment (marciw-da) [reply docs/2.8.0/docs/daml-finance/tutorials/advanced-topics/instrument-modeling/date-utilities.rst:194]: Not sure about "reveals what happens"

File: docs/2.8.0/docs/daml-finance/tutorials/advanced-topics/instrument-modeling/date-utilities.rst [hunk 16]
```diff
@@ -0,0 +1,279 @@
...
+
+.. literalinclude:: ../../../src/test/daml/Daml/Finance/Util/Test/Date/Schedule.daml
+  :language: daml
+  :start-after: -- CREATE_PERIODIC_SCHEDULE_WITH_STUB_BEGIN
+  :end-before: -- CREATE_PERIODIC_SCHEDULE_WITH_STUB_END
+
+it would only have 11 months in total, which does not match with neither 3 nor 4 regular 3M periods.
+Instead, we need to define a stub period, so that the rest of the schedule is regular.
+In the example above, we did this by defining a start date for the first regular period. This
+implies a short initial stub period, in our case 2 months. We would expect the following schedule:
```
[the remark this answers is not in the document]
Comment (marciw-da) [reply docs/2.8.0/docs/daml-finance/tutorials/advanced-topics/instrument-modeling/date-utilities.rst:204]: Do you maybe mean "assumes" rather than "implies"? Not sure.

File: docs/2.8.0/docs/daml-finance/tutorials/advanced-topics/instrument-modeling/date-utilities.rst (hunk 13 above)
Comment (marciw-da) [reply docs/2.8.0/docs/daml-finance/tutorials/advanced-topics/instrument-modeling/date-utilities.rst:274]: I'd suggest adding some sort of conclusion or next steps. The page ends sort of abruptly.

**Review by marciw-da — APPROVED** [review]
LGTM with suggestions. Happy to clarify or discuss!

File: docs/2.8.0/docs/daml-finance/tutorials/advanced-topics/instrument-modeling/date-utilities.rst (hunk 16 above)
[the remark this answers is not in the document]
Comment (markus-da) [reply docs/2.8.0/docs/daml-finance/tutorials/advanced-topics/instrument-modeling/date-utilities.rst:206]: I meant that _defining a start date for the first regular period_ has the consequence a short initial stub period will be used, so that would be "implies", right?

**Comment by markus-da** [comment]
Thank you for the detailed review! I will merge this now and create a Conclusion section as part of a separate PR. I can also fix any follow-ups there.
